### PR TITLE
feat!: fractional bucketing improvements

### LINKED ISSUE #1738: [FEATURE] High-resolution fractional bucketing in flagd provider
Sub-issue of open-feature/flagd#1903. See the [parent issue](https://github.com/open-feature/flagd/issues/1903) and [ADR](https://github.com/open-feature/flagd/pull/1800) for full details.

Replace the current 100-bucket float-based fractional bucketing with integer arithmetic using `MaxInt32` (2,147,483,647) as the max weight sum:

```
bucket = (uint64(hashValue) * uint64(totalWeight)) >> 32
```

Weights must be integers and sum to at most `MaxInt32`. Java: convert hash to unsigned via `hash & 0xFFFFFFFFL`, use `long` for multiply, `>>> 32` (unsigned shift).

This is a behavioral breaking change (bucket assignments will shift), but the API surface is unchanged.

> ⚠️ **Dependency:** complete the nested fractional evaluation feature first: open-feature/java-sdk-contrib#1737

@toddbaert will coordinate the release across providers.

### PR DESCRIPTION
**Pull request by chrfwow**
## High-precision and nested fractional

This is a coordinated feature across flagd and all flagd providers.

- `fractional` now supports up to max-int32 total weight and 1/max-int32 resolution as described in the [high-precision fractional bucketing ADR](https://flagd.dev/architecture-decisions/high-precision-fractional-bucketing)
- `fractional` now supports computed (nested JSONLogic) variants and weights as described in the [fractional ADR](https://flagd.dev/architecture-decisions/fractional)
- Includes e2e tests from the [flagd-testbed v3.1.0...v3.4.0](https://github.com/open-feature/flagd-testbed/compare/v3.1.0...v3.4.0) to validate consistent fractional bucketing across all provider implementations
- :warning: Pseudorandom hash assignments will change with this release due to changes in the hashing algorithm; update flagd and your providers to ensure fractional assignment is consistent across all workloads

### Related PRs

| Component | PR |
|---|---|
| flagd | https://github.com/open-feature/flagd/pull/1909 |
| java-sdk-contrib | https://github.com/open-feature/java-sdk-contrib/pull/1740 |
| js-sdk-contrib | https://github.com/open-feature/js-sdk-contrib/pull/1501 |
| python-sdk-contrib | https://github.com/open-feature/python-sdk-contrib/pull/373 |
| dotnet-sdk-contrib | https://github.com/open-feature/dotnet-sdk-contrib/pull/622 |
| flagd-schemas | https://github.com/open-feature/flagd-schemas/pull/207 |

> **Warning**
> @toddbaert will coordinate the release across flagd and all providers. Do not merge without coordination.
> @toddbaert will update the RPC mode tests accordingly once the associated flagd PR is merged and an up-to-date flagd-testbed image is available.

---

## Original PR description

Adds high resolution fractional support

### Related Issues

Fixes #1738

### CODE REVIEW

File: tools/flagd-core/src/test/resources/fractional/selfContainedFractionalB.json [hunk 1]
```diff
@@ -10,5 +10,5 @@
       50
     ]
   ],
-  "result": "blue"
+  "result": "red"
```
Comment (chrfwow) [inline tools/flagd-core/src/test/resources/fractional/selfContainedFractionalB.json:13]: In some cases, we will compute a different bucket now

**Comment by chrfwow** [comment]
Do we want more tests for this? I stuck to the current test suite

File: tools/flagd-core/src/main/java/dev/openfeature/contrib/tools/flagd/core/targeting/Fractional.java [hunk 2]
```diff
@@ -71,16 +81,55 @@ public Object evaluate(List arguments, Object data, String jsonPath) throws Json
...
+                    (byte) ((short) number)
+            };
+        } else if (number instanceof Float) {
+            return numberToByteArray(Float.floatToIntBits((Float) number));
+        } else if (number instanceof BigDecimal) {
+            return numberToByteArray(Double.doubleToLongBits(number.doubleValue()));
+        } else {
+            throw new IllegalArgumentException("Unsupported number type: " + number.getClass());
+        }
+    }
```
[the remark this answers is not in the document]
Comment (chrfwow) [reply tools/flagd-core/src/main/java/dev/openfeature/contrib/tools/flagd/core/targeting/Fractional.java]: Disagree with point 1. Other implementation will also convert large numbers to a Double equivalent type, so we will be consistent.
As for point 2, do we want the java.nio.Bytebuffer?

**Comment by chrfwow** [comment]
Idk why there are PMD errors, but I think they are unrelated and false positives

File: tools/flagd-core/src/test/java/dev/openfeature/contrib/tools/flagd/core/targeting/FractionalTest.java [hunk 3]
```diff
@@ -47,6 +62,67 @@ void validate_emptyJson_targetingReturned(@ConvertWith(FileContentConverter.clas
...
+    @ValueSource(ints = {
+            0,
+            1,
+            -1,
+            Integer.MAX_VALUE,
+            Integer.MAX_VALUE - 1,
+            Integer.MIN_VALUE,
+            Integer.MIN_VALUE + 1
+    })
+    void edgeCasesDoNotThrow(int hash) throws JsonLogicException {
```
Comment (chrfwow) [inline tools/flagd-core/src/test/java/dev/openfeature/contrib/tools/flagd/core/targeting/FractionalTest.java]: This only tests that we will find a bucket for edge case inputs, but it will not validate the correctness

File: tools/flagd-core/src/test/java/dev/openfeature/contrib/tools/flagd/core/targeting/FractionalTest.java [hunk 4]
```diff
@@ -47,6 +62,67 @@ void validate_emptyJson_targetingReturned(@ConvertWith(FileContentConverter.clas
...
+
+        AtomicReference<String> result = new AtomicReference<>();
+        assertDoesNotThrow(() -> result.set(Fractional.distributeValueFromHash(hash, bucketsList, totalWeight, "")));
+
+        assertNotNull(result.get());
+        assertTrue(result.get().startsWith("bucket"));
+    }
+
+    @Test
+    void statistics() throws JsonLogicException {
```
Comment (chrfwow) [inline tools/flagd-core/src/test/java/dev/openfeature/contrib/tools/flagd/core/targeting/FractionalTest.java]: This tests the statistical distribution. The test takes 340ms on my machine, I think we can live with it, but I can also delete or disable it if we want to

Comment (chrfwow) [reply tools/flagd-core/src/test/java/dev/openfeature/contrib/tools/flagd/core/targeting/FractionalTest.java]: removed

File: tools/flagd-core/src/test/java/dev/openfeature/contrib/tools/flagd/core/targeting/FractionalTest.java (hunk 3 above)
Comment (chrfwow) [reply tools/flagd-core/src/test/java/dev/openfeature/contrib/tools/flagd/core/targeting/FractionalTest.java]: removed, validated through the gherkin suite

File: tools/flagd-core/src/main/resources/flagd/schemas/flags.json [hunk 5]
```diff
@@ -1,73 +1,123 @@
 {
   "$id": "https://flagd.dev/schema/v0/flags.json",
   "$schema": "http://json-schema.org/draft-07/schema#",
-  "title": "flagd Flag Configuration",
-  "description": "Defines flags for use in flagd, including typed variants and rules.",
-  "type": "object",
-  "properties": {
-    "flags": {
+  "$ref": "#/definitions/providerConfig",
```
Comment (chrfwow) [inline tools/flagd-core/src/main/resources/flagd/schemas/flags.json:4]: I don't understand why there are changes in this file. They seem to be auto-generated when running the flagd-core test target

**Comment by chrfwow** [comment]
Waiting for https://github.com/open-feature/flagd-testbed/pull/340

File: tools/flagd-core/src/main/resources/flagd/schemas/flags.json (hunk 5 above)
Comment (aepfli) [reply tools/flagd-core/src/main/resources/flagd/schemas/flags.json:4]: could be that those are copied from the flagd-schema repo

Comment (toddbaert) [reply tools/flagd-core/src/main/resources/flagd/schemas/flags.json:4]: Ya I think that's what it is - I believe it's coming from there.

I think maybe we should gitignore it?

File: tools/flagd-core/src/main/java/dev/openfeature/contrib/tools/flagd/core/targeting/Fractional.java [hunk 6]
```diff
@@ -71,29 +80,60 @@ public Object evaluate(List arguments, Object data, String jsonPath) throws Json
...
+
+    static String distributeValueFromHash(
+            final int hash, final List<FractionProperty> propertyList, final int totalWeight, final String jsonPath)
             throws JsonLogicEvaluationException {
-        byte[] bytes = hashKey.getBytes(StandardCharsets.UTF_8);
-        int mmrHash = MurmurHash3.hash32x86(bytes, 0, bytes.length, 0);
-        float bucket = Math.abs(mmrHash) * 1.0f / Integer.MAX_VALUE * 100;
+        long longHash = Integer.toUnsignedLong(hash);
+        int bucket = Math.abs((int) ((longHash * totalWeight) >> 32));
```
Comment (toddbaert) [inline tools/flagd-core/src/main/java/dev/openfeature/contrib/tools/flagd/core/targeting/Fractional.java]: I think this `math.abs` here is misleading. Since longHash is in `[0, 2^32)`, and totalWeight is in `[0, MaxInt32]`, I think the product is always non-negative.

File: tools/flagd-core/src/main/resources/flagd/schemas/flags.json (hunk 5 above)
Comment (chrfwow) [reply tools/flagd-core/src/main/resources/flagd/schemas/flags.json:4]: it already is gitignored 🤔

File: tools/flagd-core/src/main/java/dev/openfeature/contrib/tools/flagd/core/targeting/Fractional.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (toddbaert) [reply tools/flagd-core/src/main/java/dev/openfeature/contrib/tools/flagd/core/targeting/Fractional.java]: No longer relevant.

**Comment by toddbaert** [comment]
@chrfwow there was some e2e test failures but I fixed them by updating to the latest git submodule in the testkit module, and adding the v1 exclusion to the file mode suite, and excluding v2 in the RPC suite (since flagd doesn't support this yet).

I will coordinate the release of flagd, then update the RPC suite here and release everything simultaneously.

**Comment by toddbaert** [comment]
Here's the flagd implementation: https://github.com/open-feature/flagd/pull/1909

@chrfwow Up to you if you want to create a new PR based on this branch, but it's probably easier to do this and [this](https://github.com/open-feature/java-sdk-contrib/issues/1737) in one go, as I did above.

**Comment by chrfwow** [comment]
@toddbaert I will implement the nested fractional evaluation in a different PR

File: providers/flagd/src/test/java/dev/openfeature/contrib/providers/flagd/e2e/RunRpcTest.java [hunk 7]
```diff
@@ -28,7 +28,7 @@
 @ConfigurationParameter(key = GLUE_PROPERTY_NAME, value = "dev.openfeature.contrib.providers.flagd.e2e.steps")
 @ConfigurationParameter(key = OBJECT_FACTORY_PROPERTY_NAME, value = "io.cucumber.picocontainer.PicoFactory")
 @IncludeTags({"rpc"})
-@ExcludeTags({"unixsocket", "deprecated"})
+@ExcludeTags({"unixsocket", "fractional-v2", "deprecated"})
```
Comment (toddbaert) [inline providers/flagd/src/test/java/dev/openfeature/contrib/providers/flagd/e2e/RunRpcTest.java]: Leaving this as a blocking comment; do not resolve for now. We will update flagd in concert with the providers and release at roughly the same time (at that point we'll be able to remove this "v2" exclude)

**Review by toddbaert — APPROVED** [review]
This PR looks good. Leaving this unresolved conversation to block for now though until implementations are consistent: https://github.com/open-feature/java-sdk-contrib/pull/1740/changes#r2974095951

@chrfwow related:

https://github.com/open-feature/flagd/pull/1909
https://github.com/open-feature/flagd-schemas/pull/207

**Review by aepfli — APPROVED** [review]

**Comment by toddbaert** [comment]
@chrfwow I've added the nested fractional support to this PR as well, since it's not much more effort and we've done these in tandem in the other languages: https://github.com/open-feature/java-sdk-contrib/pull/1740/commits/ffda346ab26e3015ddfe6d09b5edfb5914e72d26
